src/Promitor.Core.Scraping/Configuration/Serialization/v1/Core/MetricDefinitionDeserializer.cs

```diff
@@ -50,7 +51,7 @@ private void DeserializeMetrics(YamlMappingNode node, MetricDefinitionV1 metricD
             if (node.Children.TryGetValue(ResourcesTag, out var metricsNode))
             {
                 var resourceDeserializer = _azureResourceDeserializerFactory.GetDeserializerFor(metricDefinition.ResourceType.Value);
-                metricDefinition.Resources = resourceDeserializer.Deserialize((YamlSequenceNode)metricsNode, errorReporter);
+                metricDefinition.Resources = resourceDeserializer.Deserialize((YamlSequenceNode)metricsNode, errorReporter).ToList();
```

Comment:
Should we keep nulls into account?
```suggestion
                metricDefinition.Resources = resourceDeserializer.Deserialize((YamlSequenceNode)metricsNode, errorReporter)?.ToList();
```

Reply:
I looked at this again, and I came to the conclusion that the result can never be null (as the code currently stands). It calls https://github.com/adamconnelly/promitor/blob/validation-improvements/src/Promitor.Core.Scraping/Configuration/Serialization/Deserializer.cs#L61, which always returns a list. Because of this, I wasn't keen on adding null-propagation here because I figure that if the intention is that it should never be null, we're better knowing about it by a NullReferenceException being thrown, rather than the problem being hidden and pushed further up the stack.

However, I also took a look at `MetricDefinitionV1` and realised that we can just change the type of the collection to `IReadOnlyCollection<>`, and it allows us to remove the `ToList()` call, meaning that this isn't an issue anymore.

Just FYI - the reason I ended up changing the type from `List<>` to `IReadOnlyCollection<>` was because of the change to `IDeserializer<>` to make it covariant - that meant that I had to use an interface as the return type.

Reply:
Good idea, thanks man!